File: 446. Arithmetic Slices II - Subsequence.java
class Solution {
    public int numberOfArithmeticSlices(int[] A) {
        	if (A.length < 3) {
			return 0;
		}
		int ans = 0;
		HashMap<Integer, Integer>[] map = new HashMap[A.length];
		for (int i = 0; i < A.length; i++) {
			map[i] = new HashMap<>(i);
			for (int j = 0; j < i; j++) {
				long diff = (long) A[i] - A[j];
				if (diff <= Integer.MIN_VALUE || diff > Integer.MAX_VALUE)
					continue;
				int d = (int) diff;
				int c1 = map[i].getOrDefault(d, 0);
				int c2 = map[j].getOrDefault(d, 0);
				ans += c2;
				map[i].put(d, c1 + c2 + 1);
			}
		}

		return ans;
    }
}

File: 446. Arithmetic Slices II - Subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {     
        vector<unordered_map<long long int,long long int>>dp(nums.size());
        long long int answer=0;       
        for(int i=1;i<nums.size();i++)
        {
            long long int currele=nums[i];
            for(int j=i-1;j>=0;j--)
            {
                long long int diff=currele-nums[j]; 
                
                if(dp[j].find(diff)==dp[j].end())
                {
                    answer+=0;
                    dp[i][diff]=dp[i][diff]+1;
                }
                else
                {
                    answer+=dp[j][diff];
                    dp[i][diff]=dp[i][diff]+1+dp[j][diff];
                }
            }
        }
        return answer;
    }
};
